File: 프로그래머스/2/87946. 피로도/피로도.java
class Solution {
    static boolean[] visited;
    static int dlength;
    static int maxCount;
    public int solution(int k, int[][] dungeons) {
        dlength = dungeons.length;
        visited = new boolean[dlength];
        dfs(0, 0, k, visited, dungeons);
        
        return maxCount;
    }
    
    public static void dfs(int i, int answer, int k, boolean[] visited, int[][] dungeons){
        
        // 현재 탐험한 던전 수를 최댓값과 비교
        maxCount = Math.max(maxCount, answer);
        
        for(int j=0; j<dlength; j++){
            if(!visited[j] && k >= dungeons[j][0]){
                visited[j] = true;
                dfs(j, answer+1, k-dungeons[j][1], visited, dungeons);
                visited[j] = false;
            }
        }
    }
}

File: 프로그래머스/2/87946. 피로도/피로도.py
from itertools import permutations

def solution(k, dungeons):
    answer = -1
    dungeons = list(permutations(dungeons))
    for dungeon in dungeons:
        now_p = k # 현재 피로도
        count = 0 # 탐험 진행한 던전 수
        for d in dungeon:
            # 최소 필요 피로도, 소모 피로도
            min_p, consume_p = d
            # 최소 필요 피로도가 현재 피로도보다 크고
            # 소모 피로도가 현재 피로도보다 작을 경우 탐험 진행
            if now_p >= min_p and now_p >= consume_p:
                # 탐험 진행
                now_p -= consume_p
                count += 1
            else:
                answer = max(answer, count)
                break
        answer = max(answer, count)
    return answer